LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/CupsResponse.cs

```diff
@@ -77,12 +72,48 @@ private int WriteUriConditionally(ConcentratorCredentialType endpointType, Span<
             }
             else
             {
-                currentPosition += WriteToSpan(0, response[currentPosition..]);
+                currentPosition += WriteToSpan((byte)0, response[currentPosition..]);
             }
 
             return currentPosition;
         }
 
+        private async Task<(int, int, Stream?)> WriteFirmwareUpgradeConditionallyAsync(Memory<byte> response, int currentPosition, CancellationToken token)
+        {
+            var firmwareMismatch = !string.Equals(this.cupsUpdateInfoRequest.Package, this.cupsTwinInfo.Package, StringComparison.OrdinalIgnoreCase);
+
+            if (firmwareMismatch)
+            {
+                if (!this.cupsUpdateInfoRequest.KeyChecksums.Any(c => c == this.cupsTwinInfo.FwKeyChecksum))
+                {
+                    throw new InvalidOperationException("Remote firmware signature generated with a key whose checksum is not available in CUPS request.");
+                }
+
+                var signature = Convert.FromBase64String(this.cupsTwinInfo.FwSignature);
+                currentPosition += WriteToSpan(signature.Length + 4, response.Span[currentPosition..]);
```

Comment:
What's the reason for adding 4 to the signature length?

Reply:
https://github.com/lorabasics/basicstation/issues/145

This is the reason @atifaziz

Reply:
We should definitely reference that issue (for future readers/maintainers) as a comment in the code.

Reply:
So if I understand correctly, the `sigLen` includes the 4 bytes of `keyCRC` _and_ `sig` fields of the CUPS response, which makes sense since if `sigLen` is 0 then `keyCRC` should be skipped. So `keyCRC` and `sig` should _both_ be present or absent.